crates/fsrt/src/forge_project.rs

```diff
@@ -0,0 +1,197 @@
+        self.funcs.extend(iter.into_iter().filter_map(|entrypoint| {
+            let (func_name, path) = entrypoint.function.into_func_path();
+            let module = self.ctx.modid_from_path(&path)?;
+            let def_id = self.env.module_export(module, func_name)?;
+            Some(ResolvedEntryPoint {
+                func_name,
+                path,
+                module,
+                def_id,
+                invokable: entrypoint.invokable,
+                webtrigger: entrypoint.web_trigger,
+                admin: entrypoint.admin,
+            })
+        }));
+    }
+}
+
+#[allow(dead_code)]
+impl MockForgeProject<'_> {
+    pub fn files_from_string(string: String) -> Self {
+        let forge_manifest = ForgeManifest::create_manifest_with_func_mod(FunctionMod {
+            key: "main",
+            handler: "index.run",
+            providers: None,
+        });
+
+        let mut mock_forge_project = MockForgeProject {
+            files_name_to_source: HashMap::new(),
+            test_manifest: forge_manifest,
+        };
+
+        let different_files = string
+            .split("//")
+            .map(|f| f.replace("//", "").trim().to_string())
+            .filter(|file| !file.is_empty());
+        for file in different_files {
+            let (file_name, file_source) = file.split_once('\n').unwrap();
+            mock_forge_project
+                .files_name_to_source
+                .insert(PathBuf::from(file_name.trim()), file_source.to_string());
+        }
+
+        mock_forge_project
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ForgeProjectFromDir {
+    #[allow(dead_code)]
+    pub dir: PathBuf,
+    pub manifest_file_content: String,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct MockForgeProject<'a> {
+    pub files_name_to_source: HashMap<PathBuf, String>,
+    pub test_manifest: ForgeManifest<'a>,
+}
+
+impl<'a> ForgeProjectTrait<'a> for ForgeProjectFromDir {
+    fn load_file(&self, path: impl AsRef<Path>, sourcemap: Arc<SourceMap>) -> Arc<SourceFile> {
+        sourcemap.load_file(path.as_ref()).unwrap()
+    }
+
+    fn get_paths(&self) -> HashSet<PathBuf> {
+        collect_sourcefiles(self.dir.join("src/")).collect::<HashSet<_>>()
+    }
+
+    fn get_secret_packages(&self) -> Vec<PackageData> {
+        if let Ok(f) = std::fs::File::open("secretdata.yaml") {
+            let scrape_config: Vec<PackageData> =
+                serde_yaml::from_reader(f).expect("Failed to deserialize package");
+            scrape_config
+        } else {
+            vec![]
+        }
+    }
+
+    fn get_manifest(&self) -> ForgeManifest<'_> {
+        let out = serde_yaml::from_str(&self.manifest_file_content);
+        out.unwrap_or_default()
+    }
+}
+
+impl<'a> ForgeProjectTrait<'a> for MockForgeProject<'a> {
+    fn load_file(&self, p: impl AsRef<Path>, _: Arc<SourceMap>) -> Arc<SourceFile> {
+        let cm: Lrc<SourceMap> = Default::default();
```

Comment:
You should move the sourcemap to the `MockForgeProject` struct, so you don't have have to keep creating it here. Also, looking at this, you should probably call `self.cm.new_source_file(..)` in the test setup whenever adding a mock file, and store the `Arc<SourceFile>` in a `HashMap<PathBuf, Arc<SourceFile>>`. That way, whenever this method is called, you can just do `self.files_name_to_source.get(p.as_ref()).unwrap().clone()`.

Reply:
Fixed. Thank you :)

Reply:
But, please check to make sure I did this correctly